{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c75f2b3-26d6-4bc6-91a2-9bc898333639",
   "metadata": {},
   "outputs": [],
   "source": [
    "ENV[\"PYTHON\"] = \"\" \n",
    "using Pkg\n",
    "Pkg.build(\"PyCall\")  # Rebuild PyCall to use the internal Python\n",
    "using Revise, PyPlot, DelimitedFiles, Genie\n",
    "ENV[\"PYTHON\"] = \"\" \n",
    "using Pkg\n",
    "Pkg.build(\"PyCall\")  # Rebuild PyCall to use the internal Python\n",
    "using PyPlot, DelimitedFiles\n",
    "using Revise, Genie, DelimitedFiles, DCAUtils, JLD2, PyPlot, Statistics, LinearAlgebra\n",
    "import KitMSA: fasta2matrix, matrix2fasta\n",
    "\n",
    "\n",
    "function encode_codons_to_nuc_int(codon_seq::Vector{String})\n",
    "    # Mapping nucleotides to Integers\n",
    "    nuc_map = Dict('A'=>1, 'C'=>2, 'G'=>3, 'T'=>4, '-'=>5, 'N'=>5)\n",
    "    \n",
    "    encoded = Int[]\n",
    "    \n",
    "    for codon in codon_seq\n",
    "        for char in codon\n",
    "            # Convert each character of the codon string\n",
    "            push!(encoded, get(nuc_map, uppercase(char), 5))\n",
    "        end\n",
    "    end\n",
    "    return encoded\n",
    "end\n",
    "\n",
    "\n",
    "function read_dna_to_int(filename::String)\n",
    "    sequences = Int[]\n",
    "    current_seq = \"\"\n",
    "    n_seqs = 0\n",
    "    \n",
    "    # Dizionario di codifica\n",
    "    encoding = Dict('A'=>1, 'C'=>2, 'G'=>3, 'T'=>4, 'a'=>1, 'c'=>2, 'g'=>3, 't'=>4, '-'=>5)\n",
    "\n",
    "    lines = readlines(filename)\n",
    "    \n",
    "    # Estraiamo solo le sequenze ignorando gli header\n",
    "    seqs_list = String[]\n",
    "    temp_seq = \"\"\n",
    "    for line in lines\n",
    "        if startswith(line, \">\")\n",
    "            if !isempty(temp_seq)\n",
    "                push!(seqs_list, temp_seq)\n",
    "                temp_seq = \"\"\n",
    "            end\n",
    "        else\n",
    "            temp_seq *= strip(line)\n",
    "        end\n",
    "    end\n",
    "    push!(seqs_list, temp_seq) # Ultima sequenza\n",
    "\n",
    "    # Conversione in Matrice L x M\n",
    "    L = length(seqs_list)\n",
    "    M = length(seqs_list[1])\n",
    "    msa_int = zeros(Int, L, M)\n",
    "\n",
    "    for i in 1:L\n",
    "        for j in 1:M\n",
    "            char = seqs_list[i][j]\n",
    "            msa_int[i, j] = get(encoding, char, 5) # Default al gap se carattere ignoto\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return Int8.(msa_int')\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "wt_name = \"TEM\"\n",
    "file_wt = \"../data_directed_evolution/TEM1_pfam_DNA.fasta\"\n",
    "file_model = \"../model_natural_PSE.dat\"\n",
    "\n",
    "# Read model parameters\n",
    "println(\"Reading model parameters\")\n",
    "if wt_name == \"AAC\" || wt_name == \"DHFR\"\n",
    "        h_tmp, J_tmp = read_par_BM_0gapsave(file_model) \n",
    "else\n",
    "        h_tmp, J_tmp = read_par_BM_lettersave(file_model)\n",
    "end\n",
    "h = set_max_field_to_0save(h_tmp);\n",
    "J_tmp2 = symmetrize_Jsave(J_tmp); \n",
    "J = permutedims(J_tmp2, [1,3,2,4]);\n",
    "\n",
    "\n",
    "# Reading sequences in the seed\n",
    "println(\"Reading sequences in the seed\")\n",
    "wt = join(readdlm(file_wt, skipstart = 1));\n",
    "L_big = round(Int,length(wt)/3)\n",
    "if wt_name == \"AAC\" || wt_name == \"DHFR\"\n",
    "        wt_DNA_seq = [wt[((i-1)*3 +1):(i*3)] for i in 1:L_big];\n",
    "else\n",
    "        wt_DNA_seq = [wt[((i-1)*3 +1):(i*3)] for i in 3:L_big-1];\n",
    "end\n",
    "\n",
    "wt_seq = encode_codons_to_nuc_int(wt_DNA_seq)\n",
    "wt_amino_seq = [cod2amino[x] for x in wt_DNA_seq]\n",
    "\n",
    "Nseqs = 10^4\n",
    "rounds = 8\n",
    "mu = 0.014\n",
    "mu_bind = 0.0\n",
    "Npairs = 100\n",
    "temp = 0.0\n",
    "\n",
    "@time res = Genie.run_dir_evol_nucleo(wt_DNA_seq, Nseqs, h, J;\n",
    "                   rounds = rounds,\n",
    "                   seq_reads = Nseqs,\n",
    "                   temp = temp,  \n",
    "                   mu = mu,  \n",
    "\t\t   mu_bind = mu_bind, \n",
    "\t\t   neutral = true);\n",
    "\n",
    "@time res_repl = Genie.run_dir_evol_nucleo(wt_DNA_seq, Nseqs, h, J;\n",
    "                   rounds = rounds,\n",
    "                   seq_reads = Nseqs,\n",
    "                   temp = temp,  \n",
    "                   mu = mu, \n",
    "\t\t   mu_bind = mu_bind, \n",
    "\t\t   neutral = true);\n",
    "\n",
    "Genie.dna2fasta(\"../tem_neutral.fa\", res.final_msa_dna)\n",
    "Genie.dna2fasta(\"../tem_neutral_repl.fa\", res_repl.final_msa_dna)\n",
    "\n",
    "\n",
    "msa_dna = read_dna_to_int(\"../tem_neutral.fa\")\n",
    "msa_dna_repl = read_dna_to_int(\"../tem_neutral_repl.fa\")\n",
    "\n",
    "@time f1, _ = compute_weighted_frequencies(msa_dna, 6, 0.); f1 = reshape(f1, 5, 597);\n",
    "@time f1_repl, _ = compute_weighted_frequencies(msa_dna_repl, 6, 0.); f1_repl = reshape(f1_repl, 5, 597);\n",
    "\n",
    "for (i, aa) in enumerate(wt_seq)\n",
    "    f1[aa, i] = NaN # Mask Wild-Type\n",
    "    f1_repl[aa, i] = NaN # Mask Wild-Type\n",
    "end\n",
    "f = filter(isfinite, vec(f1))\n",
    "f_repl = filter(isfinite, vec(f1_repl))\n",
    "\n",
    "off = 1e-9;\n",
    "close(\"all\"); \n",
    "plt.scatter(f .+ off, f_repl .+ off, s=4, alpha=0.6) # Identity Line\n",
    "plt.plot([1e-6, 1], [1e-6, 1], color=\"red\", linestyle=\"--\", linewidth=0.8, alpha=0.7)\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "savefig(\"../tem_repl_freqs_dna.png\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0c97647-8530-4890-ae7e-6ef44ae30f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "ENV[\"PYTHON\"] = \"\" \n",
    "using Pkg\n",
    "Pkg.build(\"PyCall\")  # Rebuild PyCall to use the internal Python\n",
    "using Revise, PyPlot, DelimitedFiles, Genie\n",
    "ENV[\"PYTHON\"] = \"\" \n",
    "using Pkg\n",
    "Pkg.build(\"PyCall\")  # Rebuild PyCall to use the internal Python\n",
    "using PyPlot, DelimitedFiles\n",
    "using Revise, Genie, DelimitedFiles, DCAUtils, JLD2, PyPlot, Statistics, LinearAlgebra\n",
    "import KitMSA: fasta2matrix, matrix2fasta\n",
    "\n",
    "\n",
    "function encode_codons_to_nuc_int(codon_seq::Vector{String})\n",
    "    # Mapping nucleotides to Integers\n",
    "    nuc_map = Dict('A'=>1, 'C'=>2, 'G'=>3, 'T'=>4, '-'=>5, 'N'=>5)\n",
    "    \n",
    "    encoded = Int[]\n",
    "    \n",
    "    for codon in codon_seq\n",
    "        for char in codon\n",
    "            # Convert each character of the codon string\n",
    "            push!(encoded, get(nuc_map, uppercase(char), 5))\n",
    "        end\n",
    "    end\n",
    "    return encoded\n",
    "end\n",
    "\n",
    "\n",
    "function read_dna_to_int(filename::String)\n",
    "    sequences = Int[]\n",
    "    current_seq = \"\"\n",
    "    n_seqs = 0\n",
    "    \n",
    "    # Dizionario di codifica\n",
    "    encoding = Dict('A'=>1, 'C'=>2, 'G'=>3, 'T'=>4, 'a'=>1, 'c'=>2, 'g'=>3, 't'=>4, '-'=>5)\n",
    "\n",
    "    lines = readlines(filename)\n",
    "    \n",
    "    # Estraiamo solo le sequenze ignorando gli header\n",
    "    seqs_list = String[]\n",
    "    temp_seq = \"\"\n",
    "    for line in lines\n",
    "        if startswith(line, \">\")\n",
    "            if !isempty(temp_seq)\n",
    "                push!(seqs_list, temp_seq)\n",
    "                temp_seq = \"\"\n",
    "            end\n",
    "        else\n",
    "            temp_seq *= strip(line)\n",
    "        end\n",
    "    end\n",
    "    push!(seqs_list, temp_seq) # Ultima sequenza\n",
    "\n",
    "    # Conversione in Matrice L x M\n",
    "    L = length(seqs_list)\n",
    "    M = length(seqs_list[1])\n",
    "    msa_int = zeros(Int, L, M)\n",
    "\n",
    "    for i in 1:L\n",
    "        for j in 1:M\n",
    "            char = seqs_list[i][j]\n",
    "            msa_int[i, j] = get(encoding, char, 5) # Default al gap se carattere ignoto\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return Int8.(msa_int')\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "wt_name = \"TEM\"\n",
    "file_wt = \"../data_directed_evolution/TEM1_pfam_DNA.fasta\"\n",
    "file_model = \"../model_natural_PSE.dat\"\n",
    "\n",
    "# Read model parameters\n",
    "println(\"Reading model parameters\")\n",
    "if wt_name == \"AAC\" || wt_name == \"DHFR\"\n",
    "        h_tmp, J_tmp = read_par_BM_0gapsave(file_model) \n",
    "else\n",
    "        h_tmp, J_tmp = read_par_BM_lettersave(file_model)\n",
    "end\n",
    "h = set_max_field_to_0save(h_tmp);\n",
    "J_tmp2 = symmetrize_Jsave(J_tmp); \n",
    "J = permutedims(J_tmp2, [1,3,2,4]);\n",
    "\n",
    "\n",
    "#file_wt = \"../data_directed_evolution/PSE1_pfam_DNA.fasta\"\n",
    "\n",
    "# Reading sequences in the seed\n",
    "println(\"Reading sequences in the seed\")\n",
    "wt = join(readdlm(file_wt, skipstart = 1));\n",
    "L_big = round(Int,length(wt)/3)\n",
    "if wt_name == \"AAC\" || wt_name == \"DHFR\"\n",
    "        wt_DNA_seq = [wt[((i-1)*3 +1):(i*3)] for i in 1:L_big];\n",
    "else\n",
    "        wt_DNA_seq = [wt[((i-1)*3 +1):(i*3)] for i in 3:L_big-1];\n",
    "end\n",
    "\n",
    "wt_seq = encode_codons_to_nuc_int(wt_DNA_seq)\n",
    "wt_amino_seq = [cod2amino[x] for x in wt_DNA_seq]\n",
    "\n",
    "Nseqs = 10^4\n",
    "\n",
    "#for Nseqs in [10^3, 10^3, 10^5]\n",
    "rounds = 8\n",
    "mu = 0.014\n",
    "mu_bind = 0.0\n",
    "Npairs = 100\n",
    "temp = 0.0\n",
    "\n",
    "@time res = Genie.run_neutral_evol_nucleo(wt_DNA_seq, Nseqs;\n",
    "                   rounds = rounds, \n",
    "                   mu = mu);\n",
    "\n",
    "\n",
    "@time res_repl = Genie.run_neutral_evol_nucleo(wt_DNA_seq, Nseqs;\n",
    "                   rounds = rounds, \n",
    "                   mu = mu);\n",
    "\n",
    "\n",
    "Genie.dna2fasta(\"../tem_neutral_new_$(Nseqs).fa\", res.final_msa_dna)\n",
    "Genie.dna2fasta(\"../tem_neutral_repl_new_$(Nseqs).fa\", res_repl.final_msa_dna)\n",
    "\n",
    "msa_dna = read_dna_to_int(\"../tem_neutral_new_$(Nseqs).fa\")\n",
    "msa_dna_repl = read_dna_to_int(\"../tem_neutral_repl_new_$(Nseqs).fa\")\n",
    "\n",
    "@time f1, _ = compute_weighted_frequencies(msa_dna, 6, 0.); f1 = reshape(f1, 5, 597);\n",
    "@time f1_repl, _ = compute_weighted_frequencies(msa_dna_repl, 6, 0.); f1_repl = reshape(f1_repl, 5, 597);\n",
    "\n",
    "for (i, aa) in enumerate(wt_seq)\n",
    "    f1[aa, i] = NaN # Mask Wild-Type\n",
    "    f1_repl[aa, i] = NaN # Mask Wild-Type\n",
    "end\n",
    "f = filter(isfinite, vec(f1))\n",
    "f_repl = filter(isfinite, vec(f1_repl))\n",
    "\n",
    "off = 0.;\n",
    "\n",
    "close(\"all\"); \n",
    "plt.scatter(f .+ off, f_repl .+ off, s=4, alpha=0.6) # Identity Line\n",
    "plt.plot([1e-6, 1], [1e-6, 1], color=\"red\", linestyle=\"--\", linewidth=0.8, alpha=0.7)\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "savefig(\"../new_tem_repl_freqs_dna_$(Nseqs).png\")\n",
    "#end\n",
    "\n",
    "distances = [];\n",
    "for i in 1:length(wt_DNA_seq)\n",
    "    push!(distances,  mean([3-sum(collect(wt_DNA_seq[i]) .== collect(stop_codons[j])) for j in 1:3]))\n",
    "end\n",
    "\n",
    "stretched = repeat(distances, inner=3)\n",
    "\n",
    "color_matrix = zeros(5, 597); \n",
    "for i in 1:5 \n",
    "    color_matrix[i,:] .= stretched \n",
    "end\n",
    "\n",
    "for (i, aa) in enumerate(wt_seq)\n",
    "    color_matrix[aa, i] = NaN # Mask Wild-Type\n",
    "end \n",
    "\n",
    "color_filt = filter(isfinite, vec(color_matrix))\n",
    "\n",
    "off = 0.;\n",
    "\n",
    "close(\"all\"); \n",
    "plt.scatter(f .+ off, f_repl .+ off, s=40, c = color_filt, alpha=0.6) # Identity Line\n",
    "plt.plot([1e-4, 1], [1e-4, 1], color=\"red\", linestyle=\"--\", linewidth=0.8, alpha=0.7)\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "plt.colorbar()\n",
    "savefig(\"../new_pse_repl_freqs_dna_$(Nseqs).png\")\n",
    "\n",
    "#end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7527a675-596a-4779-955d-d9aa0c490243",
   "metadata": {},
   "outputs": [],
   "source": [
    "Nseqs = 10^3\n",
    "msa_dna_small = read_dna_to_int(\"../tem_neutral_new_$(Nseqs).fa\")\n",
    "msa_dna_repl_small = read_dna_to_int(\"../tem_neutral_repl_new_$(Nseqs).fa\")\n",
    "\n",
    "@time f1_small, _ = compute_weighted_frequencies(msa_dna_small, 6, 0.); f1_small = reshape(f1_small, 5, 597);\n",
    "@time f1_repl_small, _ = compute_weighted_frequencies(msa_dna_repl_small, 6, 0.); f1_repl_small = reshape(f1_repl_small, 5, 597);\n",
    "\n",
    "for (i, aa) in enumerate(wt_seq)\n",
    "    f1_small[aa, i] = NaN # Mask Wild-Type\n",
    "    f1_repl_small[aa, i] = NaN # Mask Wild-Type\n",
    "end\n",
    "f = filter(isfinite, vec(f1_small))\n",
    "f_repl = filter(isfinite, vec(f1_repl_small))\n",
    "\n",
    "\n",
    "close(\"all\"); \n",
    "plt.hist(f1[:], histtype = \"step\", label = \"10^5\");\n",
    "plt.hist(f1_small[:], histtype = \"step\", label = \"10^3\"); \n",
    "plt.xlabel(\"freq\"); \n",
    "plt.legend();\n",
    "savefig(\"../ciao.png\")\n",
    "\n",
    "\n",
    "\n",
    "using StatsBase\n",
    "using PyPlot\n",
    "\n",
    "idx = findall(x -> 0.001 < x < 0.018, f1)\n",
    "\n",
    "#idx = findall(x -> x > 0.01, f1)\n",
    "# ----------------------------\n",
    "# Estrai lettere e siti\n",
    "# ----------------------------\n",
    "letters = getindex.(idx, 1)   # 1..5\n",
    "sites   = getindex.(idx, 2)    # 1..597\n",
    "\n",
    "# ----------------------------\n",
    "# Parametri\n",
    "# ----------------------------\n",
    "q = 5\n",
    "L = 597\n",
    "\n",
    "# ----------------------------\n",
    "# Profili di abbondanza\n",
    "# ----------------------------\n",
    "count_letters = countmap(letters)\n",
    "count_sites   = countmap(sites)\n",
    "\n",
    "letters_profile = [get(count_letters, l, 0) for l in 1:q]\n",
    "sites_profile   = [get(count_sites,   s, 0) for s in 1:L]\n",
    "\n",
    "# ----------------------------\n",
    "# Plot\n",
    "# ----------------------------\n",
    "figure(figsize=(12,4))\n",
    "\n",
    "# Lettere\n",
    "subplot(1,2,1)\n",
    "bar(1:q, letters_profile)\n",
    "title(\"aminoacids\")\n",
    "xticks(1:q)\n",
    "\n",
    "# Siti\n",
    "subplot(1,2,2)\n",
    "plot(1:L, sites_profile, linewidth=2)\n",
    "title(\"Sites\")\n",
    "\n",
    "tight_layout()\n",
    "\n",
    "# ----------------------------\n",
    "# Salva figura\n",
    "# ----------------------------\n",
    "savefig(\"../profilo_lettere_siti.png\", dpi=300, bbox_inches=\"tight\")\n",
    "\n",
    "show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b779f692-c798-4a30-a997-1df702013037",
   "metadata": {},
   "outputs": [],
   "source": [
    "Nseqs = 3+10^5\n",
    "\n",
    "rounds = 8\n",
    "mu = 0.014\n",
    "mu_bind = 0.0\n",
    "Npairs = 100\n",
    "temp = 0.0\n",
    "\n",
    "#wt_DNA_seq2 = [\"AAA\" for i in 1:199]\n",
    "#wt_seq2 = [1 for i in 1:597]\n",
    "@time res = Genie.run_neutral_evol_nucleo_nofilt(wt_DNA_seq, Nseqs;\n",
    "                   rounds = rounds, \n",
    "                   mu = mu);\n",
    "\n",
    "\n",
    "@time res_repl = Genie.run_neutral_evol_nucleo_nofilt(wt_DNA_seq, Nseqs;\n",
    "                   rounds = rounds, \n",
    "                   mu = mu);\n",
    "\n",
    "\n",
    "Genie.dna2fasta(\"../TEM_neutral_new_$(Nseqs).fa\", res.final_msa_dna)\n",
    "Genie.dna2fasta(\"../TEM_neutral_repl_new_$(Nseqs).fa\", res_repl.final_msa_dna)\n",
    "\n",
    "msa_dna = read_dna_to_int(\"../TEM_neutral_new_$(Nseqs).fa\")\n",
    "msa_dna_repl = read_dna_to_int(\"../TEM_neutral_repl_new_$(Nseqs).fa\")\n",
    "\n",
    "@time f1, _ = compute_weighted_frequencies(msa_dna, 6, 0.); f1 = reshape(f1, 5, 597);\n",
    "@time f1_repl, _ = compute_weighted_frequencies(msa_dna_repl, 6, 0.); f1_repl = reshape(f1_repl, 5, 597);\n",
    "\n",
    "for (i, aa) in enumerate(wt_seq)\n",
    "    f1[aa, i] = NaN # Mask Wild-Type\n",
    "    f1_repl[aa, i] = NaN # Mask Wild-Type\n",
    "end\n",
    "f = filter(isfinite, vec(f1))\n",
    "f_repl = filter(isfinite, vec(f1_repl))\n",
    "\n",
    "off = 0.;\n",
    "\n",
    "close(\"all\"); \n",
    "plt.scatter(f .+ off, f_repl .+ off, s=4, alpha=0.6) # Identity Line\n",
    "#plt.plot([1e-6, 1], [1e-6, 1], color=\"red\", linestyle=\"--\", linewidth=0.8, alpha=0.7)\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "savefig(\"../new_TEM_repl_freqs_dna_$(Nseqs)_nofilt.png\")\n",
    "#end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.0",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
