{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5975a163-80d9-47b1-83b9-f5cab8c2b714",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "h = rand(21,202)\n",
    "J = rand(21,202,21,202)\n",
    "\n",
    "@time res = run_evolution(Int8.(rand(1:21,(,1000))), h, J, p = 0, N_steps = 100, N_points = 10);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c9f48ab-8544-4864-bd48-2b4b74870e6d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "548a5215-6c10-4691-b0b4-8427e7cca171",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Revise, Genie, PyPlot, Statistics, Random, JLD2, Distributions, StatsBase, DCAUtils, DelimitedFiles, BenchmarkTools, LinearAlgebra, PyPlot\n",
    "import KitMSA: fasta2matrix, letter2num, num2letter, extract_params, read_par_BM, set_max_field_to_0, symmetrize_J\n",
    "\n",
    "h, J = extract_params(\"../scra_data_beta/Parameters_conv_Matteo_pc.dat\"; q = 21);\n",
    "J = permutedims(J, [1,3,2,4]);\n",
    "\n",
    "L = 202; q = 21; N_chains = 2; rng = random_gens(N_chains); temp = 1.; no_cod_bias = \n",
    "Dict(x => (1/length(amino2cod[cod2amino[x]])) for x in keys(cod2amino)); codon_usage = no_cod_bias;\n",
    "chains = [Chain(Int8.(rand(1:q,L)), q, rng[n]) for n in 1:N_chains]; codon_net = Genie.create_nested_codon_dict(); length_of_moves = create_length_dict(codon_net);\n",
    "seq_site = rand(chains[1].generator, findall(x -> x != 21, chains[1].seq)); nucleo_site = rand(chains[1].generator,1:3);\n",
    "\n",
    "length_of_moves[chains[1].DNA[seq_site], nucleo_site]\n",
    "codon_list = accessible_codons(chains[1].DNA[seq_site], codon_net, nucleo_site)\n",
    "log_prob = zeros(size(codon_list,1))\n",
    "\n",
    "@code_warntype Genie.accessible_codons_old(chains[1].DNA[seq_site], codon_net, nucleo_site)\n",
    "@code_warntype Genie.accessible_codons(chains[1].DNA[seq_site], codon_net, nucleo_site)\n",
    "\n",
    "@btime log_prob!($(log_prob), $(chains[1]), $(h), $(J), $(codon_usage), $(codon_list), $(temp), $(seq_site), $(L))\n",
    "@code_warntype log_prob!(log_prob, chains[1], h, J, codon_usage, codon_list, temp, seq_site, L)\n",
    "\n",
    "@btime prob_cond!($(chains[1]), $(h), $(J), $(seq_site), $(nucleo_site), $(codon_net), $(codon_usage), $(length_of_moves),$(temp), $(L))\n",
    "@code_warntype prob_cond!(chains[1], h, J, seq_site, nucleo_site, codon_net, codon_usage, length_of_moves, temp, L)\n",
    "\n",
    "\n",
    "#@btime prob_cond2!($(chains[1]), $(h), $(J), $(seq_site), $(nucleo_site), $(codon_net), $(codon_usage), $(temp), $(L))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33349ee5-befe-4067-921d-4006958f020f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## test proposal of metropolis\n",
    "\n",
    "target = 63/64\n",
    "n_it = 10^4\n",
    "all_codons = vcat([amino2cod[i] for i in 1:20]...)\n",
    "for cod in all_codons\n",
    "    res = []\n",
    "    for i in 1:n_it\n",
    "        push!(res,Genie.del_ins_codon_sampling(Genie.get_accessible_nucleo_for_del_ins(cod)[2]))\n",
    "    end\n",
    "    \n",
    "    println(\"Old method: Codon $(cod) difference from target $(sum(res .== cod)/n_it  - target)\")\n",
    "end\n",
    "\n",
    "\n",
    "target = 63/64\n",
    "n_it = 10^4\n",
    "all_codons = vcat([amino2cod[i] for i in 1:20]...)\n",
    "for cod in all_codons\n",
    "    res = []\n",
    "    for i in 1:n_it\n",
    "        push!(res, Genie.proposed_codon(cod, all_codons))\n",
    "    end\n",
    "    \n",
    "    println(\"New method: Codon $(cod) difference from target $(sum(res .== nothing)/n_it - target)\")\n",
    "end\n",
    "\n",
    "target = 1/64\n",
    "n_it = 10^5\n",
    "all_codons = vcat([amino2cod[i] for i in 1:20]...)\n",
    "res = [Genie.del_ins_codon_sampling(Genie.get_accessible_nucleo_for_del_ins(\"---\")[2]) for _ in 1:n_it]\n",
    "\n",
    "for cod in all_codons\n",
    "    println(\"Old method: Codon $(cod) difference from target $(sum(res .== cod)/n_it  - target)\")\n",
    "end\n",
    "\n",
    "\n",
    "target = 1/64\n",
    "n_it = 10^5\n",
    "all_codons = vcat([amino2cod[i] for i in 1:20]...)\n",
    "res = [Genie.proposed_codon(cod, all_codons) for _ in 1:n_it]\n",
    "\n",
    "for cod in all_codons\n",
    "    println(\"New method: Codon $(cod) difference from target $(sum(res .== cod)/n_it  - target)\")\n",
    "end\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1811ee28-338b-4fb5-8fc8-53a04b54cfe5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## short term test\n",
    "\n",
    "here i test against two experiments, for tem the correlation should be 0.78, for pse around 0.86\n",
    "the final hamming for tem should be 17/18, for pse 23/24\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de4e7f73-073c-4b59-8768-7c71ad18a4f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Revise, Genie, PyPlot, Statistics, Random, JLD2, Distributions, StatsBase, DCAUtils, DelimitedFiles, BenchmarkTools, LinearAlgebra, PyPlot\n",
    "import KitMSA: fasta2matrix, letter2num, num2letter, extract_params, read_par_BM, set_max_field_to_0, symmetrize_J\n",
    "\n",
    "h, J = extract_params(\"../scra_data_beta/Parameters_conv_Matteo_pc.dat\"; q = 21);\n",
    "J = permutedims(J, [1,3,2,4]);\n",
    "\n",
    "path_wt_tem = \"../Gen.jl/data/alignments/refs/TEM1_pfam_DNA.fasta\"; path_wt_pse = \n",
    "\"../Gen.jl/data/alignments/refs/PSE1_pfam_DNA.fasta\"; wt_tem = join(readdlm(path_wt_tem, \n",
    "        skipstart = 1)); wt_pse = join(readdlm(path_wt_pse, skipstart = 1)); L_tem = \n",
    "Int64(length(wt_tem)/3); tem_DNA_seq = \n",
    "[wt_tem[((i-1)*3 +1):(i*3)] for i in 1:L_tem]; tem_amino_seq = [cod2amino[codon] \n",
    "    for codon in tem_DNA_seq]; L_pse = Int64(length(wt_pse)/3); pse_DNA_seq = \n",
    "[wt_pse[((i-1)*3 +1):(i*3)] for i in 1:L_pse]; pse_amino_seq = [cod2amino[codon] \n",
    "    for codon in pse_DNA_seq];\n",
    "\n",
    "path_fantini = \"../Gen.jl/data/alignments/experimental/fantini.max.6.gaps.gen12.fa\"; MSA_tem_exp = \n",
    "Int8.(fasta2matrix(path_fantini)); f1_tem,f2_tem = \n",
    "DCAUtils.compute_weighted_frequencies(Int8.(MSA_tem_exp'), 22, 0.); conn_tem = f2_tem .- f1_tem' * f1_tem;\n",
    "\n",
    "MSA_pse_exp = Int8.(fasta2matrix(\"../scra_data_beta/cleaned_aligned_stiffler_MSA_big.fasta\")); f1_pse,f2_pse = \n",
    "DCAUtils.compute_weighted_frequencies(Int8.(MSA_pse_exp'), 22, 0.); conn_pse = f2_pse .- f1_pse' * f1_pse;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8f4b21-239e-4eef-807d-1ae3eafa5439",
   "metadata": {},
   "outputs": [],
   "source": [
    "##test with fantini tem\n",
    "start_msa = hcat([tem_DNA_seq for i in 1:size(MSA_tem_exp,1)]...); @time res_tem = run_evolution(start_msa, \n",
    "    h, \n",
    "    J,\n",
    "    p = 0., \n",
    "    temp = 1.14, \n",
    "    N_steps = 109,  \n",
    "    N_points = 20);\n",
    "\n",
    "f1_sil_tem, f2_sil_tem = DCAUtils.compute_weighted_frequencies(Int8.(res_tem.step_msa[end]), 22, 0.); conn_sil_tem = f2_sil_tem .- \n",
    "f1_sil_tem' * f1_sil_tem;\n",
    "\n",
    "println(\"\"); println(\"Final tem sil hamming : $(mean(ham_dist(tem_amino_seq, Int8.(res_tem.step_msa[end])))) +- $(std(ham_dist(tem_amino_seq, Int8.(res_tem.step_msa[end]))))\"); println(\n",
    "    \"Final tem exp hamming : $(mean(ham_dist(tem_amino_seq, Int8.(MSA_tem_exp')))) +- $(std(ham_dist(tem_amino_seq, Int8.(MSA_tem_exp'))))\")  \n",
    "\n",
    "println(\"\"); println(\"Final tem sil energy : $(mean(energy(Int8.(res_tem.step_msa[end]), h, J))) +- $(std(energy(Int8.(res_tem.step_msa[end]), h, J)))\"); println(\n",
    "    \"Final tem exp energy : $(mean(energy(Int8.(MSA_tem_exp'), h, J))) +- $(std(energy(Int8.(MSA_tem_exp'), h, J)))\")  \n",
    "\n",
    "println(\"\"); println(\"1-point tem correlation $(corspearman(f1_tem[:], f1_sil_tem[:]))\")\n",
    "\n",
    "close(\"all\"); plt.plot(f1_tem[:], f1_sil_tem[:], \"o\"); plt.xscale(\"log\"); plt.yscale(\"log\"); savefig(\"../prova_tem.png\")\n",
    "\n",
    "\n",
    "#test with stiffler pse\n",
    "start_msa = hcat([pse_DNA_seq for i in 1:size(MSA_pse_exp,1)]...); @time res = run_evolution(start_msa, \n",
    "    h, \n",
    "    J,\n",
    "    p = 0., \n",
    "    temp = 1.4, \n",
    "    N_steps = 120,  \n",
    "    N_points = 20);\n",
    "\n",
    "f1_sil_pse, f2_sil_pse = DCAUtils.compute_weighted_frequencies(Int8.(res.step_msa[end]), \n",
    "    22, 0.); conn_sil_pse = f2_sil_pse .- f1_sil_pse' * f1_sil_pse;\n",
    "\n",
    "println(\"\"); println(\"1-point correlation pse $(corspearman(f1_pse[:], f1_sil_pse[:]))\")\n",
    "\n",
    "println(\"\"); println(\"Final pse sil hamming : $(mean(ham_dist(pse_amino_seq, Int8.(res.step_msa[end])))) +- $(std(ham_dist(pse_amino_seq, Int8.(res.step_msa[end]))))\"); println(\n",
    "    \"Final pse exp hamming : $(mean(ham_dist(pse_amino_seq, Int8.(MSA_pse_exp')))) +- $(std(ham_dist(pse_amino_seq, Int8.(MSA_pse_exp'))))\")  \n",
    "\n",
    "println(\"\"); println(\"Final pse sil energy : $(mean(energy(Int8.(res.step_msa[end]), h, J))) +- $(std(energy(Int8.(res.step_msa[end]), h, J)))\"); println(\n",
    "    \"Final pse exp energy : $(mean(energy(Int8.(MSA_pse_exp'), h, J))) +- $(std(energy(Int8.(MSA_pse_exp'), h, J)))\")  \n",
    "\n",
    "\n",
    "close(\"all\"); plt.plot(f1_pse[:], f1_sil_pse[:], \"o\"); plt.xscale(\"log\"); plt.yscale(\"log\"); savefig(\"../prova_pse.png\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc9841db-4313-4537-832a-50381c6ad803",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Long term test\n",
    "\n",
    "\n",
    "here i test the generativity for the dbd family\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d69c29-ba31-4085-9118-61d61f298f1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "@load \"../data_Genie/pars_dbd.jld2\"\n",
    "\n",
    "dbd_msa  = read_fasta_alignment(\"../Gen.jl/data/alignments/natural/DBD_alignment.uniref90.cov80.a2m\", 0.9);\n",
    "\n",
    "wt_thornton = Int8.(readdlm(\"../Gen.jl/data/alignments/refs/wt_thornton\")[:]);\n",
    "start_msa = hcat([dbd_msa[:,1] for i in 1:1000]...); @time res = run_evolution(start_msa, \n",
    "    h_dbd, \n",
    "    J_dbd,\n",
    "    p = 0.1, \n",
    "    temp = 1., \n",
    "    N_steps = 600000,  \n",
    "    N_points = 20, \n",
    "    verbose = true);\n",
    "\n",
    "check_equilibration(\"../figures/dbd_p0.1/\", dbd_msa, res.step_msa, res.steps)\n",
    "\n",
    "Genie.check_timescales(res.step_msa, \n",
    "    dbd_msa, h_dbd, J_dbd, res.steps, \"../figures/dbd/dynamics.png\");\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc089361-b1e4-4b21-a044-659b24a568c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "h, J = extract_params(\"../scra_data_beta/Parameters_conv_Matteo_pc.dat\"; q = 21);\n",
    "J = permutedims(J, [1,3,2,4]);\n",
    "\n",
    "\n",
    "\n",
    "beta_msa  = read_fasta_alignment(\"../Gen.jl/data/alignments/natural/PF13354_noinsert_max19gaps_nodupl_noclose.faa\", 0.9);\n",
    "\n",
    "start_msa = hcat([pse_DNA_seq for i in 1:1000]...); @time res = run_evolution(start_msa, \n",
    "    h, \n",
    "    J,\n",
    "    p = 0.5, \n",
    "    temp = 1., \n",
    "    N_steps = 10^7,  \n",
    "    N_points = 20, \n",
    "    verbose = true);\n",
    "\n",
    "check_equilibration(\"../figures/beta/\", beta_msa, res.step_msa, res.steps)\n",
    "\n",
    "Genie.check_timescales(res.step_msa, \n",
    "    beta_msa, h, J, res.steps, \"../figures/beta/dynamics.png\");\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ebc05e1-8338-48cf-90a4-6f6c2d6838fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_msa = hcat([wt_hinds for i in 1:1000]...); @time res = run_evolution(start_msa, \n",
    "           h, \n",
    "           J,\n",
    "           p = 0.5, \n",
    "           temp = 1., \n",
    "           N_steps = 4000000,  \n",
    "           N_points = 150, \n",
    "           verbose = true); @time check_equilibration(\"../figures/dhfr/\", \n",
    "    dhfr_msa, res.step_msa); @time Genie.check_timescales(res.step_msa, \n",
    "    dhfr_msa, h, J, res.steps, \"../figures/dhfr/dynamics.png\");\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.0",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
