{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7479b23-8141-4699-bee8-7d529b05af4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "using Revise, Genie, DelimitedFiles, DCAUtils, JLD2, PyPlot, KitMSA, Statistics\n",
    "\n",
    "@load \"../data_Genie/pars_beta_nc.jld2\"\n",
    "\n",
    "\n",
    "beta_msa  = read_fasta_alignment(\"../Gen.jl/data/alignments/natural/PF13354_noinsert_max19gaps_nodupl_noclose.faa\", 0.9);\n",
    "\n",
    "\n",
    "#define paths\n",
    "path_wt_tem = \"../Gen.jl/data/alignments/refs/TEM1_pfam_DNA.fasta\";\n",
    "path_wt_pse = \"../Gen.jl/data/alignments/refs/PSE1_pfam_DNA.fasta\";\n",
    "\n",
    "#load wt\n",
    "wt_tem = join(readdlm(path_wt_tem, skipstart = 1));\n",
    "wt_pse = join(readdlm(path_wt_pse, skipstart = 1));\n",
    "\n",
    "#transform wt_dna in amino\n",
    "L_tem = Int64(length(wt_tem)/3);\n",
    "tem_DNA_seq = [wt_tem[((i-1)*3 +1):(i*3)] for i in 1:L_tem];\n",
    "tem = Int.([Genie.cod2amino[codon] for codon in tem_DNA_seq]);\n",
    "\n",
    "L_pse = Int64(length(wt_pse)/3);\n",
    "pse_DNA_seq = [wt_pse[((i-1)*3 +1):(i*3)] for i in 1:L_pse];\n",
    "pse = Int.([Genie.cod2amino[codon] for codon in pse_DNA_seq]);\n",
    "\n",
    "\n",
    "N_start = 10000; @time m = Genie.run_dir_evol(pse[3:end-1], N_start, h[:,3:end-1], J[:,3:end-1,:,3:end-1];\n",
    "                   rounds = 4,\n",
    "                   seq_reads = 1000,\n",
    "                   each_step = 1,\n",
    "                   temp = 1.,  \n",
    "                   mu = 10^-2, mu_bind = 18.6);\n",
    "\n",
    "\n",
    "mean(ham_dist(pse[3:end-1], m.step_msa[end]))\n",
    "\n",
    "\n",
    "N_start = 10000; @time m_n = Genie.run_dir_evol_nucleo(pse_DNA_seq[3:end-1], N_start, h[:,3:end-1], J[:,3:end-1,:,3:end-1];\n",
    "                   rounds = 4,\n",
    "                   seq_reads = 1000,\n",
    "                   each_step = 1,\n",
    "                   temp = 1.,  \n",
    "                   mu = 10^-2, mu_bind = 18.6);\n",
    "\n",
    "\n",
    "mean(ham_dist(pse[3:end-1], m_n.step_msa[end]))\n",
    "\n",
    "\n",
    "function give_pos(idx)\n",
    "    seq_site = div(idx,3) + 1\n",
    "    nucleo_site = mod(idx,3)\n",
    "            if seq_site == 0\n",
    "                seq_site = 1\n",
    "            end\n",
    "            if nucleo_site == 0\n",
    "                nucleo_site = 3\n",
    "            end\n",
    "    return seq_site, nucleo_site\n",
    "end\n",
    "\n",
    "start_msa = hcat([beta_msa[:,1] for i in 1:10000]...); @time res = run_evolution(start_msa, \n",
    "    h, \n",
    "    J,\n",
    "    p = 0.5, \n",
    "    temp = 1., \n",
    "    N_steps = 5*10^2,  \n",
    "    N_points = 20, \n",
    "    verbose = true);\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "M = 1000000; temp = 1; mu_bind = 18.6; mu = 10^-1;  start_msa = Int.(hcat([pse for i in 1:M]...)); L = size(start_msa,1 );msa = deepcopy(\n",
    "    start_msa); \n",
    "@time Genie.random_mutations!(msa, mu);\n",
    "#@time abund_dict = Genie.seq_count_dict(msa);\n",
    "@time Genie.selection!(msa, h, J, temp = 1., mu_bind = 18.6);\n",
    "final_msa = zeros(Int, L, M); @time Genie.amplification!(final_msa, msa);\n",
    "\n",
    "mean(ham_dist(Int.(pse), final_msa))\n",
    "\n",
    "\n",
    "\n",
    "M = 10000; temp = 1.; mu_bind = 18.6; mu = 10^-1;  start_msa = Int.(hcat([pse for i in 1:M]...)); L = size(start_msa,1);msa = deepcopy(\n",
    "    start_msa); \n",
    "@btime Genie.random_mutations!($(msa), $(mu));\n",
    "@btime abund_dict = Genie.seq_count_dict($(msa));\n",
    "@btime Genie.selection!($(msa), $(abund_dict), $(h), $(J), temp = temp, mu_bind = mu_bind);\n",
    "final_msa = zeros(Int, L, M); @btime Genie.amplification!($(final_msa), $(msa), $(abund_dict));\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "check_equilibration(\"../figures/beta/\", beta_msa, res.step_msa, res.steps)\n",
    "\n",
    "Genie.check_timescales(res.step_msa, \n",
    "    beta_msa, h, J, res.steps, \"../figures/beta/dynamics.png\");\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.0",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
